README.md
[![Build Status](https://travis-ci.org/mapbox/tilelive-multicache.png?branch=master)](https://travis-ci.org/mapbox/tilelive-multicache)

tilelive-multicache
--------------
Module for adding a caching layer in front a tilelive source. It wraps a tilelive backend providing a new source constructor with cached superpowers!

    var options = {
        client: client, // Required, instantiated client
        ttl: 300,       // optional, object cache ttl in seconds
        stale: 300      // optional, max number of seconds to allow a stale object to be served
    };
    var TileJSON = require('tilelive-multicache')(options, require('tilejson'));

    new TileJSON( ... )


The `client` option must be an object with the following methods;

- `set(key, ttl, value, callback)`
- `get(key, callback)`
- `error(err)`


CHANGELOG.md
## 0.0.2

- Fixes an issue with error handling

## 0.0.1

- Initial release, based on [tilelive-memcached](https://github.com/mapbox/tilelive-memcached) and [tilelive-redis](https://github.com/mapbox/tilelive-redis)


test/unit.test.js
var assert = require('assert');
var tap = require('tap');
var TileliveCache = require('../index');
var bufferEqual = require('buffer-equal');

var test = tap.test;

test(' encode', function(t) {
    var errstatCode404 = new Error(); errstatCode404.statusCode = 404;
    var errstatCode403 = new Error(); errstatCode403.statusCode = 403;
    var errstatCode500 = new Error(); errstatCode500.statusCode = 500;
    assert.equal(TileliveCache.encode(errstatCode404), '404');
    assert.equal(TileliveCache.encode(errstatCode403), '403');
    assert.equal(TileliveCache.encode(errstatCode500), null);

    assert.ok(bufferEqual(TileliveCache.encode(null, {id:'foo'}), new Buffer(
        '{"x-tl-json":true}' +
        new Array(1025 - '{"x-tl-json":true}'.length).join(' ') +
        '{"id":"foo"}'
    )), 'encodes object');

    assert.ok(bufferEqual(TileliveCache.encode(null, 'hello world'), new Buffer(
        '{}' +
        new Array(1025 - '{}'.length).join(' ') +
        'hello world'
    ), 'encodes string'));

    assert.ok(bufferEqual(TileliveCache.encode(null, new Buffer(0)), new Buffer(
        '{}' +
        new Array(1025 - '{}'.length).join(' ') +
        ''
    ), 'encodes empty buffer'));

    assert.ok(bufferEqual(TileliveCache.encode(null, new Buffer(0), { 'content-type': 'image/png' }), new Buffer(
        '{"content-type":"image/png"}' +
        new Array(1025 - '{"content-type":"image/png"}'.length).join(' ') +
        ''
    ), 'encodes headers'));

    assert.throws(function() {
        TileliveCache.encode(null, new Buffer(0), { data: new Array(1024).join(' ') });
    }, Error, 'throws when headers exceed 1024 bytes');

    t.end();
});

test('decode', function(t) {
    assert.deepEqual(TileliveCache.decode('404'), {err:{statusCode:404,tlcache:true}});
    assert.deepEqual(TileliveCache.decode('403'), {err:{statusCode:403,tlcache:true}});

    var headers = JSON.stringify({'x-tl-json':true,'x-tl-cache':'hit'});
    var encoded = new Buffer(
        headers +
        new Array(1025 - headers.length).join(' ') +
        JSON.stringify({'id':'foo'})
    );
    assert.deepEqual(TileliveCache.decode(encoded), {
        headers:{'x-tl-json':true,'x-tl-cache':'hit'},
        buffer:{'id':'foo'}
    }, 'decodes object');

    headers = JSON.stringify({'x-tl-cache':'hit'});
    encoded = new Buffer(
        headers +
        new Array(1025 - headers.length).join(' ') +
        'hello world'
    );
    assert.deepEqual(TileliveCache.decode(encoded), {
        headers:{'x-tl-cache':'hit'},
        buffer: new Buffer('hello world')
    }, 'decodes string (as buffer)');

    headers = JSON.stringify({'x-tl-cache':'hit'});
    encoded = new Buffer(
        headers +
        new Array(1025 - headers.length).join(' ') +
        ''
    );
    assert.deepEqual(TileliveCache.decode(encoded), {
        headers:{'x-tl-cache':'hit'},
        buffer: new Buffer(0)
    }, 'decodes empty buffer');

    headers = JSON.stringify({'content-length':'5'});
    encoded = new Buffer(
        headers +
        new Array(1025 - headers.length).join(' ') +
        '0'
    );
    assert.throws(function() {
        TileliveCache.decode(encoded);
    }, Error, 'throws on bad content-length');

    headers = JSON.stringify({'content-length':'1'});
    encoded = new Buffer(
        headers +
        new Array(1025 - headers.length).join(' ') +
        '0'
    );
    assert.doesNotThrow(function() {
        TileliveCache.decode(encoded);
    }, Error, 'matches good content-length');

    encoded = new Buffer('bogus');
    assert.throws(function() {
        TileliveCache.decode(encoded);
    }, Error, 'throws when encoded buffer does not include headers');

    t.end();
});


test/cache.test.js
var assert = require('assert');
var tap = require('tap');
var LRU = require('lru-cache');
var TileliveCache = require('../index');

var Testsource = require('./testsource');

var tiles = Testsource.tiles;
var grids = Testsource.grids;
var now = Testsource.now;

var test = tap.test;

function lruClient() {
    var cache = LRU();

    return {
        get: function(k, cb) {
            var v = cache.get(k);
            cb(null, v);
        },
        set: function(k, t, v, cb) {
            cache.set(k, v, t * 100);
            cb(null);
        },
        error: function(err) {
            console.error(err); // eslint-disable-line no-console
        }
    };
}

test('load', function(t) {

    t.test('fails without source', function(t) {
        assert.throws(function() { TileliveCache({}); });
        assert.throws(function() { TileliveCache({}, {}); });
        t.end();
    });
    t.test('loads + sets default values', function(t) {
        var Source = TileliveCache({ client: lruClient() }, Testsource);
        assert.ok(Source.options);
        assert.ok(Source.options.ttl, 300);
        new Source('fakeuri', function(err, source) {
            assert.ifError(err);
            assert.ok(source instanceof Testsource);
            assert.equal(source._uri, 'fakeuri');
            t.end();
        });
    });
    t.test('sets config from opts', function(t) {
        var Source = TileliveCache({
            client: lruClient(),
            ttl: 10,
            stale: 5
        }, Testsource);
        assert.ok(Source.options);
        assert.equal(Source.options.ttl, 10);
        assert.equal(Source.options.stale, 5);
        t.end();
    });

    t.end();
});

var tile = function(expected, cached, done) {
    return function(err, data, headers) {
        assert.ifError(err);
        assert.ok(data instanceof Buffer);
        assert.ok(cached ? headers['x-tl-cache'] : !headers['x-tl-cache']);
        assert[cached ? 'deepEqual' : 'strictEqual'](data, expected);
        assert.equal(data.length, expected.length);
        assert.equal(headers['content-type'], 'image/png');
        assert.equal(headers['last-modified'], now.toUTCString());
        done();
    };
};
var grid = function(expected, cached, done) {
    return function(err, data, headers) {
        assert.ifError(err);
        assert.ok(cached ? headers['x-tl-cache'] : !headers['x-tl-cache']);
        assert.deepEqual(data, expected);
        assert.equal(headers['content-type'], 'application/json');
        assert.equal(headers['last-modified'], now.toUTCString());
        done();
    };
};
var error = function(message, cached, done) {
    return function(err /*, data, headers */) {
        assert.ok(cached ? err.tlcache: !err.tlcache);
        assert.equal(err.message, message);
        done();
    };
};

test('source', function(t) {
    var source;
    t.test('create source', function(t) {
        var CachedSource = TileliveCache({
            client: lruClient(),
            ttl: 1,
            stale: 1
        }, Testsource);
        source = new CachedSource({}, function(err) {
            assert.ifError(err);
            t.end();
        });
    });

    t.test('tile 200 a miss', function(t) {
        source.getTile(0, 0, 0, tile(tiles.a, false, t.end));
    });
    t.test('tile 200 a hit', function(t) {
        source.getTile(0, 0, 0, tile(tiles.a, true, t.end));
    });
    t.test('tile 200 b miss', function(t) {
        source.getTile(1, 0, 0, tile(tiles.b, false, t.end));
    });
    t.test('tile 200 b hit', function(t) {
        source.getTile(1, 0, 0, tile(tiles.b, true, t.end));
    });
    t.test('tile 40x miss', function(t) {
        source.getTile(4, 0, 0, error('Tile does not exist', false, t.end));
    });
    t.test('tile 40x hit', function(t) {
        source.getTile(4, 0, 0, error('Tile does not exist', true, t.end));
    });
    t.test('tile 500 miss', function(t) {
        source.getTile(2, 0, 0, error('Unexpected error', false, t.end));
    });
    t.test('tile 500 miss', function(t) {
        source.getTile(2, 0, 0, error('Unexpected error', false, t.end));
    });
    t.test('grid 200 a miss', function(t) {
        source.getGrid(0, 0, 0, grid(grids.a, false, t.end));
    });
    t.test('grid 200 a hit', function(t) {
        source.getGrid(0, 0, 0, grid(grids.a, true, t.end));
    });
    t.test('grid 200 b miss', function(t) {
        source.getGrid(1, 0, 0, grid(grids.b, false, t.end));
    });
    t.test('grid 200 b hit', function(t) {
        source.getGrid(1, 0, 0, grid(grids.b, true, t.end));
    });
    t.test('grid 40x miss', function(t) {
        source.getGrid(4, 0, 0, error('Grid does not exist', false, t.end));
    });
    t.test('grid 40x hit', function(t) {
        source.getGrid(4, 0, 0, error('Grid does not exist', true, t.end));
    });

    t.test('expire', function(t) {
        setTimeout(function() {
            t.test('tile 200 a expires', function(t) {
                source.getTile(0, 0, 0, tile(tiles.a, false, t.end));
            });
            t.test('tile 200 b expires', function(t) {
                source.getTile(1, 0, 0, tile(tiles.b, false, t.end));
            });
            t.test('tile 40x expires', function(t) {
                source.getTile(4, 0, 0, error('Tile does not exist', false, t.end));
            });
            t.test('grid 200 a expires', function(t) {
                source.getGrid(0, 0, 0, grid(grids.a, false, t.end));
            });
            t.test('grid 200 b expires', function(t) {
                source.getGrid(1, 0, 0, grid(grids.b, false, t.end));
            });
            t.test('grid 40x expires', function(t) {
                source.getGrid(4, 0, 0, error('Grid does not exist', false, t.end));
            });
            t.end();
        }, 2000);
    });
    t.end();
});

test('refresh', function(t) {
    var source;
    t.test('create source', function(t) {
        var CachedSource = TileliveCache({
            client: lruClient(),
            ttl: 1,
            stale: 300
        }, Testsource);
        source = new CachedSource({}, function(err) {
            assert.ifError(err);
            t.end();
        });
    });

    var origExpires;
    t.test('tile 200 a miss', function(t) {
        source.getTile(0, 0, 0, function(err, data, headers) {
            origExpires = headers['x-tl-expires'];
            tile(tiles.a, false, function() {
                t.end();
            })(err, data, headers);
        });
    });
    t.test('tile 200 a hit (stale)', function(t) {
        setTimeout(function() {
            source.getTile(0, 0, 0, function(err, data, headers) {
                assert.equal(origExpires, headers['x-tl-expires']);
                tile(tiles.a, true, function(){
                    t.end();
                })(err, data, headers);
            });
        }, 1000);
    });
    t.test('tile 200 a hit (fresh)', function(t) {
        setTimeout(function() {
            source.getTile(0, 0, 0, function(err, data, headers) {
                assert.notEqual(origExpires, headers['x-tl-expires']);
                tile(tiles.a, true, function(){
                    t.end();
                })(err, data, headers);
                
            });
        }, 1000);
    });
    t.end();
});
    
test('upstream expires', function(t) {
    var customExpires;
    var stats = {};
    var getter = function(id, callback) {
        stats[id] = stats[id] || 0;
        stats[id]++;

        var err;
        if (id === 'missing') {
            err = new Error('Not found');
            err.statusCode = 404;
            return callback(err);
        }
        if (id === 'fatal') {
            err = new Error('Fatal');
            err.statusCode = 500;
            return callback(err);
        }
        if (id === 'nocode') {
            err = new Error('Unexpected');
            return callback(err);
        }

        return callback(null, {id:id}, { Expires: customExpires });
    };
    var wrapped = TileliveCache.cachingGet('test', {
        client: lruClient()
    }, getter);
    customExpires = (new Date(+new Date() + 1000)).toUTCString();
    
    t.test('getter 200 miss', function(t) {
        wrapped('asdf', function(err, data, headers) {
            assert.ifError(err);
            assert.deepEqual(data, {id:'asdf'}, 'returns data');
            assert.deepEqual(headers.expires, customExpires, 'passes customExpires through');
            assert.deepEqual(headers['x-tl-expires'], customExpires, 'sets x-tl-expires based on customExpires');
            assert.deepEqual(headers['x-tl-cache'], undefined, 'cache miss');
            assert.equal(stats.asdf, 1, 'asdf IO x1');
            t.end();
        });
    });
    t.test('getter 200 hit', function(t) {
        wrapped('asdf', function(err, data, headers) {
            assert.ifError(err);
            assert.deepEqual(data, {id:'asdf'}, 'returns data');
            assert.deepEqual(headers.expires, customExpires, 'passes customExpires through');
            assert.deepEqual(headers['x-tl-expires'], customExpires, 'sets x-tl-expires based on customExpires');
            assert.deepEqual(headers['x-tl-cache'], 'hit');
            assert.deepEqual(headers['x-tl-json'], true);
            assert.equal(stats.asdf, 1, 'asdf IO x1');
            t.end();
        });
    });
    t.test('getter 200 miss', function(t) {
        setTimeout(function() {
            wrapped('asdf', function(err, data, headers) {
                assert.ifError(err);
                assert.deepEqual(data, {id:'asdf'}, 'returns data');
                assert.deepEqual(headers.expires, customExpires, 'passes customExpires through');
                assert.deepEqual(headers['x-tl-expires'], customExpires, 'sets x-tl-expires based on customExpires');
                assert.deepEqual(headers['x-tl-cache'], undefined, 'cache miss');
                assert.equal(stats.asdf, 2, 'asdf IO x2');
                t.end();
            });
        }, 3000);
    });
    t.end();
});

test('cachingGet', function(t) {
    var stats = {};
    var getter = function(id, callback) {
        stats[id] = stats[id] || 0;
        stats[id]++;

        var err;
        if (id === 'missing') {
            err = new Error('Not found');
            err.statusCode = 404;
            return callback(err);
        }
        if (id === 'fatal') {
            err = new Error('Fatal');
            err.statusCode = 500;
            return callback(err);
        }
        if (id === 'nocode') {
            err = new Error('Unexpected');
            return callback(err);
        }

        return callback(null, {id:id});
    };
    var wrapped = TileliveCache.cachingGet('test', {
        client: lruClient()
    }, getter);
    t.test('getter 200 miss', function(t) {
        wrapped('asdf', function(err, data, headers) {
            assert.ifError(err);
            assert.deepEqual(data, {id:'asdf'}, 'returns data');
            assert.deepEqual(Object.keys(headers), ['x-tl-expires', 'x-tl-json'], 'sets x-tl-expires header');
            assert.equal(stats.asdf, 1, 'asdf IO x1');
            t.end();
        });
    });
    t.test('getter 200 hit', function(t) {
        wrapped('asdf', function(err, data, headers) {
            assert.ifError(err);
            assert.deepEqual(data, {id:'asdf'}, 'returns data');
            assert.deepEqual(Object.keys(headers), ['x-tl-expires', 'x-tl-json', 'x-tl-cache'], 'sets x-tl-expires header');
            assert.deepEqual(headers['x-tl-cache'], 'hit');
            assert.deepEqual(headers['x-tl-json'], true);
            assert.equal(stats.asdf, 1, 'asdf IO x1');
            t.end();
        });
    });
    t.test('getter 404 miss', function(t) {
        wrapped('missing', function(err, data, headers) {
            assert.equal(err.toString(), 'Error: Not found', 'not found err');
            assert.equal(err.statusCode, 404, 'err statusCode 404');
            assert.deepEqual(Object.keys(headers), ['x-tl-expires'], 'sets x-tl-expires header');
            assert.equal(stats.missing, 1, 'missing IO x1');
            t.end();
        });
    });
    t.test('getter 404 hit', function(t) {
        wrapped('missing', function(err, data, headers) {
            assert.equal(err.toString(), 'Error', 'not found err');
            assert.equal(err.statusCode, 404, 'err statusCode 404');
            assert.ok(!headers, 'no headers');
            assert.equal(stats.missing, 1, 'missing IO x1');
            t.end();
        });
    });
    t.test('getter 500 miss', function(t) {
        wrapped('fatal', function(err, data, headers) {
            assert.equal(err.toString(), 'Error: Fatal', 'fatal err');
            assert.equal(err.statusCode, 500, 'err statusCode 500');
            assert.ok(!headers, 'no headers');
            assert.equal(stats.fatal, 1, 'fatal IO x1');
            t.end();
        });
    });
    t.test('getter 500 miss', function(t) {
        wrapped('fatal', function(err, data, headers) {
            assert.equal(err.toString(), 'Error: Fatal', 'fatal err');
            assert.equal(err.statusCode, 500, 'err statusCode 500');
            assert.ok(!headers, 'no headers');
            assert.equal(stats.fatal, 2, 'fatal IO x1');
            t.end();
        });
    });
    t.test('getter nocode', function(t) {
        wrapped('nocode', function(err, data, headers) {
            assert.equal(err.toString(), 'Error: Unexpected', 'unexpected err');
            assert.equal(err.statusCode, undefined, 'no err statusCode');
            assert.ok(!headers, 'no headers');
            assert.equal(stats.nocode, 1, 'nocode IO x1');
            t.end();
        });
    });
    t.test('getter nocode', function(t) {
        wrapped('nocode', function(err, data, headers) {
            assert.equal(err.toString(), 'Error: Unexpected', 'unexpected err');
            assert.equal(err.statusCode, undefined, 'no err statusCode');
            assert.ok(!headers, 'no headers');
            assert.equal(stats.nocode, 2, 'nocode IO x1');
            t.end();
        });
    });
    t.end();
});


test('perf', function(t) {
    var buffer = require('fs').readFileSync(__dirname + '/encode-buster.pbf');
    t.test('encodes buster PBF in < 10ms', function(t) {
        var time = + new Date();
        for (var i = 0; i < 10; i++) TileliveCache.encode(null, buffer);
        time = + new Date() - time;
        assert.equal(time < 10, true, 'encodes buster PBF 10x in ' + time + 'ms');
        t.end();
    });
    t.end();
});


test('perf-source', function(t) {
    var source;
    t.test('create source', function(t) {
        var CachedSource = TileliveCache({
            client: lruClient()
        }, Testsource);
        source = new CachedSource({hostname: 'perf'}, function(err) {
            assert.ifError(err);
            t.end();
        });
    });
    t.test('gets buster tile 10x in < 20ms', function(t) {
        var remaining = 10;
        var time = + new Date();
        for (var i = 0; i < 10; i++) source.getTile(0,0,0, function(err, data /*, headers */) {
            assert.ifError(err);
            assert.equal(data.length, 783167);
            if (!--remaining) {
                time = + new Date() - time;
                assert.equal(time < 40, true, 'getTile buster PBF 10x in ' + time + 'ms');
                t.end();
            }
        });
    });
    t.end();
});


test/memjs.test.js
var assert = require('assert');
var tap = require('tap');
var TileliveCache = require('../index');
var memjs = require('memjs');
var Testsource = require('./testsource');

var tiles = Testsource.tiles;
var grids = Testsource.grids;
var now = Testsource.now;

var test = tap.test;

function client(cacheClient) {
    var cache = cacheClient || memjs.Client.create(null, {'keepAlive': true});

    return {
        get: function(k, cb) {
            cache.get(k, cb);
        },
        set: function(k, t, v, cb) {
            cache.set(k, v, cb, t);
        },
        error: function(err) {
            console.error(err); // eslint-disable-line no-console
        },
        memcached: cache
    };
}
function deadClient() {
    return client(memjs.Client.create('127.0.0.1:11212'));
}

var tile = function(expected, cached, done) {
    return function(err, data, headers) {
        assert.ifError(err);
        assert.ok(data instanceof Buffer);
        assert.ok(cached ? headers['x-tl-cache'] : !headers['x-tl-cache']);
        assert[cached ? 'deepEqual' : 'strictEqual'](data, expected);
        assert.equal(data.length, expected.length);
        assert.equal(headers['content-type'], 'image/png');
        assert.equal(headers['last-modified'], now.toUTCString());
        done();
    };
};
var grid = function(expected, cached, done) {
    return function(err, data, headers) {
        assert.ifError(err);
        assert.ok(cached ? headers['x-tl-cache'] : !headers['x-tl-cache']);
        assert.deepEqual(data, expected);
        assert.equal(headers['content-type'], 'application/json');
        assert.equal(headers['last-modified'], now.toUTCString());
        done();
    };
};
var error = function(message, cached, done) {
    return function(err /*, data, headers */) {
        assert.ok(cached ? err.tlcache: !err.tlcache);
        assert.equal(err.message, message);
        done();
    };
};

test('memjs client', function(t) {
    var CachedSource = TileliveCache({
        client: client(),
        ttl: 1,
        stale: 1
    }, Testsource);
    var source;
    t.test('create source', function(t) {
        source = new CachedSource({}, function(err) {
            assert.ifError(err);
            CachedSource.options.client.memcached.flush(function() {
                t.end();
            });
        });
    });

    t.test('tile 200 a miss', function(t) {
        source.getTile(0, 0, 0, tile(tiles.a, false, t.end));
    });
    t.test('tile 200 a hit', function(t) {
        source.getTile(0, 0, 0, tile(tiles.a, true, t.end));
    });
    t.test('tile 200 b miss', function(t) {
        source.getTile(1, 0, 0, tile(tiles.b, false, t.end));
    });
    t.test('tile 200 b hit', function(t) {
        source.getTile(1, 0, 0, tile(tiles.b, true, t.end));
    });
    t.test('tile 40x miss', function(t) {
        source.getTile(4, 0, 0, error('Tile does not exist', false, t.end));
    });
    t.test('tile 40x hit', function(t) {
        source.getTile(4, 0, 0, error('Tile does not exist', true, t.end));
    });
    t.test('tile 500 miss', function(t) {
        source.getTile(2, 0, 0, error('Unexpected error', false, t.end));
    });
    t.test('tile 500 miss', function(t) {
        source.getTile(2, 0, 0, error('Unexpected error', false, t.end));
    });
    t.test('grid 200 a miss', function(t) {
        source.getGrid(0, 0, 0, grid(grids.a, false, t.end));
    });
    t.test('grid 200 a hit', function(t) {
        source.getGrid(0, 0, 0, grid(grids.a, true, t.end));
    });
    t.test('grid 200 b miss', function(t) {
        source.getGrid(1, 0, 0, grid(grids.b, false, t.end));
    });
    t.test('grid 200 b hit', function(t) {
        source.getGrid(1, 0, 0, grid(grids.b, true, t.end));
    });
    t.test('grid 40x miss', function(t) {
        source.getGrid(4, 0, 0, error('Grid does not exist', false, t.end));
    });
    t.test('grid 40x hit', function(t) {
        source.getGrid(4, 0, 0, error('Grid does not exist', true, t.end));
    });

    t.test('expire', function(t) {
        setTimeout(function() {
            t.test('tile 200 a expires', function(t) {
                source.getTile(0, 0, 0, tile(tiles.a, false, t.end));
            });
            t.test('tile 200 b expires', function(t) {
                source.getTile(1, 0, 0, tile(tiles.b, false, t.end));
            });
            t.test('tile 40x expires', function(t) {
                source.getTile(4, 0, 0, error('Tile does not exist', false, t.end));
            });
            t.test('grid 200 a expires', function(t) {
                source.getGrid(0, 0, 0, grid(grids.a, false, t.end));
            });
            t.test('grid 200 b expires', function(t) {
                source.getGrid(1, 0, 0, grid(grids.b, false, t.end));
            });
            t.test('grid 40x expires', function(t) {
                source.getGrid(4, 0, 0, error('Grid does not exist', false, t.end));
            });
            t.end();
        }, 2000);
    });

    t.test('quit', function(t) {
        CachedSource.options.client.memcached.quit();
        t.end();
    });

    t.end();
});

test('Dead memjs client', function(t) {
    var CachedSource = TileliveCache({
        client: deadClient(),
        ttl: 1,
        stale: 1
    }, Testsource);
    var source;
    t.test('create source', function(t) {
        source = new CachedSource({}, function(err) {
            assert.ifError(err);
            CachedSource.options.client.memcached.flush(function() {
                t.end();
            });
        });
    });
    t.test('dead tile 200 a miss', function(t) {
        source.getTile(0, 0, 0, tile(tiles.a, false, t.end));
    });
    t.test('dead tile 200 b miss', function(t) {
        source.getTile(1, 0, 0, tile(tiles.b, false, t.end));
    });
    t.test('dead grid 200 a miss', function(t) {
        source.getGrid(0, 0, 0, grid(grids.a, false, t.end));
    });
    t.test('dead grid 200 b miss', function(t) {
        source.getGrid(1, 0, 0, grid(grids.b, false, t.end));
    });

    t.test('quit', function(t) {
        CachedSource.options.client.memcached.quit();
        t.end();
    });

    t.end();
});


index.js
var util = require('util');

module.exports = function(options, Source) {
    if (!Source) throw new Error('No source provided');
    if (!Source.prototype.get) throw new Error('No get method found on source');

    function Caching() { return Source.apply(this, arguments); }

    // Inheritance.
    util.inherits(Caching, Source);

    // References for testing, convenience, post-call overriding.
    Caching.options = options;

    Caching.prototype.get = module.exports.cachingGet('TL4', options, Source.prototype.get);

    return Caching;
};

module.exports.cachingGet = function(namespace, options, get) {
    if (!get) throw new Error('No get function provided');
    if (!namespace) throw new Error('No namespace provided');

    options = options || {};
    options.stale = typeof options.stale === 'number' ? options.stale : 300;
    options.ttl = typeof options.ttl === 'number' ? options.ttl : 300;

    if (!options.client) throw new Error('No cache client');
    if (!options.stale) throw new Error('No stale option set');
    if (!options.ttl) throw new Error('No ttl option set');

    return function(url, callback) {
        var key = namespace + '-' + url;
        var source = this;
        var client = options.client;
        var stale = options.stale;
        var ttl = options.ttl;

        client.get(key, function(err, encoded) {
            // If error on get, pass through to original source
            // without attempting a set after retrieval.
            if (err) {
                err.key = key;
                client.error(err);
                return get.call(source,url, callback);
            }

            // Cache hit.
            var data;
            if (encoded) try {
                data = decode(encoded);
            } catch(e) {
                e.key = key;
                client.error(e);
            }
            if (data) {
                callback(data.err, data.buffer, data.headers);
                if (isFresh(data)) return;

                // Update cache & bump `expires` header
                get.call(source, url, function(err, buffer, headers) {
                    if (err && !errcode(err)) return client.error(err);

                    headers = headers || {};
                    headers = setEx(key, err, buffer, headers, ttl, stale);
                });
            } else {
                // Cache miss, error, or otherwise no data
                get.call(source, url, function(err, buffer, headers) {
                    if (err && !errcode(err)) return callback(err);

                    headers = headers || {};
                    headers = setEx(key, err, buffer, headers, ttl, stale);
                    callback(err, buffer, headers);
                });
            }
        });

        function setEx(key, err, buffer, headers, ttl, stale) {
            var expires = headers.Expires || headers.expires;
            delete headers.Expires;
            delete headers.expires;
            if (expires) {
                headers.expires = expires;
                headers['x-tl-expires'] = expires;
            } else {
                headers['x-tl-expires'] = (new Date(Date.now() + (ttl * 1000))).toUTCString();
            }

            // seconds from now to expiration time
            var sec = Math.ceil((Number(new Date(headers['x-tl-expires'])) - Number(new Date()))/1000);

            // stale is the number of extra seconds to cache an object
            // past its expires time where we may serve a "stale"
            // version of the object.
            //
            // When an upstream expires is set no stale padding is used
            // so that the upstream expires is fully respected.
            var pad = expires ? 0 : stale;

            if (sec > 0) client.set(key, sec + pad, encode(err, buffer, headers), function(err) {
                if (!err) return;
                err.key = key;
                client.error(err);
            });

            return headers;
        }

        function isFresh(d) {
            // When we don't have an expires header just assume staleness
            if (d.headers === undefined || !d.headers['x-tl-expires']) return false;

            return (+(new Date(d.headers['x-tl-expires'])) > Date.now());
        }
    };
};

module.exports.encode = encode;
module.exports.decode = decode;

function errcode(err) {
    if (!err) return;
    if (err.statusCode === 404) return 404;
    if (err.statusCode === 403) return 403;
    return;
}

function encode(err, buffer, headers) {
    if (errcode(err)) return errcode(err).toString();

    // Unhandled error.
    if (err) return null;

    headers = headers || {};

    // Turn objects into JSON string buffers.
    if (buffer && typeof buffer === 'object' && !(buffer instanceof Buffer)) {
        headers['x-tl-json'] = true;
        buffer = new Buffer(JSON.stringify(buffer));
    // Turn strings into buffers.
    } else if (buffer && !(buffer instanceof Buffer)) {
        buffer = new Buffer(buffer);
    }

    headers = new Buffer(JSON.stringify(headers), 'utf8');

    if (headers.length > 1024) {
        throw new Error('Invalid cache value - headers exceed 1024 bytes: ' + JSON.stringify(headers));
    }

    var padding = new Buffer(1024 - headers.length);
    padding.fill(' ');
    var len = headers.length + padding.length + buffer.length;
    return Buffer.concat([headers, padding, buffer], len);
}

function decode(encoded) {
    if (encoded.length == 3) {
        encoded = encoded.toString();
        if (encoded === '404' || encoded === '403') {
            var err = new Error();
            err.statusCode = parseInt(encoded, 10);
            err.tlcache= true;
            return { err: err };
        }
    }

    // First 1024 bytes reserved for header + padding.
    var offset = 1024;
    var data = {};
    data.headers = encoded.slice(0, offset).toString().trim();

    try {
        data.headers = JSON.parse(data.headers);
    } catch(e) {
        throw new Error('Invalid cache value');
    }

    data.headers['x-tl-cache'] = 'hit';
    data.buffer = encoded.slice(offset);

    // Return JSON-encoded objects to true form.
    if (data.headers['x-tl-json']) data.buffer = JSON.parse(data.buffer);

    if (data.headers['content-length'] && data.headers['content-length'] != data.buffer.length)
        throw new Error('Content length does not match');
    return data;
}
